sycl/doc/extensions/proposed/sycl_ext_intel_device_aspects.asciidoc

```diff
@@ -0,0 +1,313 @@
+= sycl_ext_intel_device_aspects
+
+:source-highlighter: coderay
+:coderay-linenums-mode: table
+
+// This section needs to be after the document title.
+:doctype: book
+:toc2:
+:toc: left
+:encoding: utf-8
+:lang: en
+:dpcpp: pass:[DPC++]
+
+// Set the default source code type in this document to C++,
+// for syntax highlighting purposes.  This is needed because
+// docbook uses c++ and html5 uses cpp.
+:language: {basebackend@docbook:c++:cpp}
+
+
+== Notice
+
+[%hardbreaks]
+Copyright (C) 2022-2022 Intel Corporation.  All rights reserved.
+
+Khronos(R) is a registered trademark and SYCL(TM) and SPIR(TM) are trademarks
+of The Khronos Group Inc.  OpenCL(TM) is a trademark of Apple Inc. used by
+permission by Khronos.
+
+
+== Contact
+
+To report problems with this extension, please open a new issue at:
+
+https://github.com/intel/llvm/issues
+
+
+== Dependencies
+
+This extension is written against the SYCL 2020 revision 5 specification.  All
+references below to the "core SYCL specification" or to section numbers in the
+SYCL specification refer to that revision.
+
+
+== Status
+
+This is a proposed extension specification, intended to gather community
+feedback.  Interfaces defined in this specification may not be implemented yet
+or may be in a preliminary state.  The specification itself may also change in
+incompatible ways before it is finalized.  *Shipping software products should
+not rely on APIs defined in this specification.*
+
+[comment]
+--
+_Add the following paragraph when this specification becomes "experimental"._
+
+There are important limitations with the DPC++ implementation of this
+experimental extension.  See the section below titled "Limitations with the
+experimental version" for a description of these limitations.
+--
+
+
+== Overview
+
+Most device aspects provide a way for an application to test whether a device
+supports a specific feature.  However, sometimes an application needs to
+conditionally execute code based on the specific model of the device.  This
+extension adds new aspects, for supported Intel devices, which can be used for
+this purpose.
+
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_INTEL_DEVICE_ASPECTS` to one of the values defined in the table
+below.  Applications can test for the existence of this macro to determine if
+the implementation supports this feature, or applications can test the macro's
+value to determine which of the extension's features the implementation
+supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|Initial version of this extension.
+|===
+
+=== New aspects
+
+This extension adds the following aspects to `sycl::aspect`:
+
+```
+namespace sycl {
+
+enum class aspect : /* unspecified */ {
+  ext_intel_x86_64,
```

Comment:
I think this sets a bad precedent for aspects: having a separate aspect for every architecture and SKU is going to lead to a very long list of aspects, and the aspects are set up such that any device can only ever have exactly one of the aspects. 

If I understand correctly, testing one of these aspects effectively tests a bunch of other aspects implicitly, so it seems pretty unlikely that anybody would ever use `if_device_has<one_of_these_new_aspects>` in conjunction with any other aspects.  For example, once you know that something is x86_64, you also know that it supports fp64, atomic64, cpu, and a bunch of other things; nobody would ever write something like `if_device_has<ext_intel_x86_64, fp64, atomic64>`.

I think things would be cleaner if we separated these architecture descriptions from aspects.  We'd need two query interfaces, but everything else could remain basically as specified. Most of the draft implementation (the macros, the new compiler switches, etc) wouldn't have to be changed -- only the definition of `if_device_has`.

That would leave us with: 
- `if_device_has<Aspects...>` => Test if a device has a specific set of fine-grained features
- `if_device_is<Architecture>` => Test if a device is a specific architecture (since it can only be one of them anyway)

This separation would enable us to teach people that aspects are fine-grained features that may be shared by multiple architectures, while architectures are more coarse-grained. That would enable developers to use aspects for future-proof code (e.g. because all Intel GPUs after a certain architecture are expected to have aspect X), and architectures when they don't care about being future-proof (i.e. because the next Intel GPU will have a different architecture name).

(I'm not tied to the name "architecture" here, but I had to pick something.)

Reply:
I think there are pros and cons to this.  Here are some more things to consider:

* I expected these aspects to also be used (eventually) in the `[[sycl::device_has()]]` C++ attribute or the proposed `device_has` kernel property.  If we change them to be a new enumeration like `architecture`, we'd have to add a new C++ attribute and / or a new kernel property named something like `device_architecture_is`.

* I could easily imagine adding designations in the future for categories of GPU architecture like "intel_gpu_xe".  Would that be added to the "aspect" enum or to the "architecture" enum?  This starts to blur the lines between the two enums, and this might be the strongest reason to represent everything as aspects.  If we do have two enums, I think "intel_gpu_xe" would probably be part of the "architecture" enum.

* Adding a new API like `if_device_architecture_is` let's us define new semantics when multiple things are listed.  Although `if_device_has` uses the "AND" semantic when there are multiple aspects, it would make more sense for `if_device_architecture_is` to use "OR" semantics.  This does seem useful, and it might be the strongest reason to have two different enums for aspects vs. architectures.

Thoughts?

Reply:
I don't think I understand the idea for `info::device::ext_intel_architecture`.  Are you saying that you want to support code like:

```
using sycl::ext::intel::experimental;

architecture a = dev.get_info<info::device::architecture>();
switch (a) {
  case architecture::intel_gpu_pvc:
  /* ... */
}
```

If that is the goal, I don't see why the info descriptor should return the type `size_t` vs. the `architecture` enumeration.

Reply:
You're right, it can return the `architecture` enumeration.  I clearly need more coffee. :)

Reply:
I think I like @Pennycook's proposal.  @kbobrovs, do you have any concerns?

Reply:
I like this too, along with the elaborations about `[[sycl::device_architecture_is()]]` attribute and ability to check architecture at runtime in host code. BTW, `target` instead of `device_architecture` could make APIs more compact.

Reply:
Time to bikeshed on the name then.  I like "device_architecture_is", which is why I used it in my examples.  The full set of APIs would be:

* `if_device_architecture_is<>()`: Selects code inside kernel.
* `device::ext_intel_architecture_is()`: Query device from host.
* `device::get_info<ext::intel::experimental::info::device::architecture>()`: Get architecture from host code.
* `device_architecture_is`: Name of kernel property.
* `[[sycl::device_architecture_is()]]`: C++ attribute (if needed).

The name "target" doesn't read well to me.  Take `if_device_target_is` ... we generally use the name "target" as a synonym of "device", so "if_device_target_is" seems redundant.  I'm not too worried that the word "architecture" is longer than "target".  This is an advanced feature that few end users will use.

Reply:
my proposal is `if_target_is`. But I can live with `if_device_architecture_is` too

Reply:
> my proposal is `if_target_is`. But I can live with `if_device_architecture_is` too

`sycl::target` is already an `enum` used for defining `accessors` (see [here](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#_interface_for_buffer_command_accessors)), so we'd need to call it something like `target_device` (which is longer than `architecture`).

If verbosity is the issue, `device_arch` maintains the link to "architecture" while being shorter and easier to spell.

Reply:
I refactored things according to this discussion in 12b0f655ed56c24dea5f50894e5dfeb7edf243a0.

I kept the longer version of the name `if_device_architecture_is` because I have a general preference to avoid abbreviations in API names.  As I type this, though, I wonder if we could drop the word "device" and shorten it to `if_architecture_is`.  I think I like that better ...

Reply:
I like `if_architecture_is` more than `if_device_architecture_is`. Other changes LGTM.

Reply:
Since we are  bikeshedding, I'll ask for another opinion.  What should be the name of the  "else" function.  We can't use "else" because it is a C++ keyword.  One option is `else_` (trailing underscore):

```
if_architecture_is<architecture::intel_gpu_bdw>([] {
  /* ... */
}).else_([] {
  /* ... */
});
```

Other possibilities are:

* `.otherwise([] { /*...*/ })`
* `.else_architecture([] { /*...*/ })`

Opinions?  Alternatives?

Reply:
Just brainstorming and not pushing for it: extra function

```
cond_architecture_is<...>([] { },  // true
                          [] { }); // false
```

Reply:
What about `or_else`? Proposed in wg21.link/p2505 for a similar case.

Reply:
@aelovikov-intel: This was actually the first iteration of this API!  We decided against it because of @AerialMantis's proposal to return an object with "else_if" and "else" member functions.  This allows us to write if-elseif-else chains of arbitrary length:

```
if_architecture_is<architecture::intel_gpu_bdw>([] {
  /* ... */
}).else_if_architecture_is<architecture::intel_gpu_skl>([] {
  /* ... */
}).else_([] {
 /* ... */
});
```

Reply:
@rolandschulz: That's not bad, but it creates a bit of an asymmetry with the `else_if_architecture_is` function:

```
if_architecture_is<architecture::intel_gpu_bdw>([] {
  /* ... */
}).else_if_architecture_is<architecture::intel_gpu_skl>([] {
  /* ... */
}).or_else([] {
 /* ... */
});
```

Another possibility is to use `fallback`:

```
if_architecture_is<architecture::intel_gpu_bdw>([] {
  /* ... */
}).else_if_architecture_is<architecture::intel_gpu_skl>([] {
  /* ... */
}).fallback([] {
 /* ... */
});
```

Reply:
> * `.otherwise([] { /*...*/ })`

I really like `otherwise`. It's the closest to the English I'd use to describe these sorts of conditions: "A if X, B if Y, and Z otherwise".

I don't like `fallback` as much.  It's a good word for the default path in this sort of device specialization, but it looks out of place to me alongside the `if`/`else`.

Reply:
Does it change your mind when you see it in the context of a long chain?

```
if_architecture_is<architecture::intel_gpu_bdw>([] {
  /* ... */
}).else_if_architecture_is<architecture::intel_gpu_skl>([] {
  /* ... */
}).otherwise([] {
 /* ... */
});
```

It seems a it weird to me to have `else_if_architecture` and then `otherwise`.  We could rename the middle function to `otherwise_if_architecture_is`, but that's a lot of typing:

```
if_architecture_is<architecture::intel_gpu_bdw>([] {
  /* ... */
}).otherwise_if_architecture_is<architecture::intel_gpu_skl>([] {
  /* ... */
}).otherwise([] {
 /* ... */
});
```

My concern with `otherwise` (and my concern with `or_else`) are both fairly weak, though.

FWIW, I kind of liked `else_` because it kept the symmetry with `else_if_architecture_is`, but I'm not very wild about the trailing underscore.

Reply:
Since it seems like discussion has stopped, I decided on `otherwise`.  See eb8aed634ff6dda0a6515fdc86517e8e99e884b9.